src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,598 @@
+                // the 'inner' encoding, then re-encode those chars under the 'this' encoding.
+                // We've already calculated the worst-case expansions for the intermediate buffers,
+                // so we use GetChars / GetBytes instead of Convert to simplify the below code
+                // and to ensure an exception is thrown if the Encoding reported an incorrect
+                // worst-case expansion.
+
+                if (_pendingReadData.Count == 0)
+                {
+                    byte[] rentedBytes = ArrayPool<byte>.Shared.Rent(DefaultReadByteBufferSize);
+                    char[] rentedChars = ArrayPool<char>.Shared.Rent(_readCharBufferMaxSize);
+
+                    try
+                    {
+                        // Beware: Use our constant value instead of 'rentedBytes.Length' when creating
+                        // the Mem<byte> struct. The reason for this is that the array pool could've returned
+                        // a larger-than-expected array, but our worst-case expansion calculations
+                        // performed earlier didn't take that into account.
+
+                        int innerBytesReadJustNow = await _innerStream.ReadAsync(rentedBytes.AsMemory(0, DefaultReadByteBufferSize), cancellationToken).ConfigureAwait(false);
+                        bool isEofReached = (innerBytesReadJustNow == 0);
+
+                        // convert bytes [inner] -> chars, then convert chars -> bytes [this]
+
+                        int charsDecodedJustNow = _innerDecoder.GetChars(rentedBytes, 0, innerBytesReadJustNow, rentedChars, 0, flush: isEofReached);
+                        int pendingReadDataPopulatedJustNow = _thisEncoder.GetBytes(rentedChars, 0, charsDecodedJustNow, _pendingReadData.Array!, 0, flush: isEofReached);
+
+                        _pendingReadData = new ArraySegment<byte>(_pendingReadData.Array!, 0, pendingReadDataPopulatedJustNow);
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(rentedBytes);
+                        ArrayPool<char>.Shared.Return(rentedChars);
+                    }
+                }
+
+                // At this point: (a) we've populated our pending read buffer and there's
+                // useful data to return to our caller; or (b) the pending read buffer is
+                // empty because the inner stream has reached EOF and all pending read data
+                // has already been flushed, and we should return 0.
+
+                int bytesToReturn = Math.Min(_pendingReadData.Count, buffer.Length);
+                _pendingReadData.AsSpan(0, bytesToReturn).CopyTo(buffer.Span);
+                _pendingReadData = _pendingReadData[bytesToReturn..];
+                return bytesToReturn;
+            }
+        }
+
+        public override int ReadByte()
+        {
+            Span<byte> buffer = stackalloc byte[1];
```

Comment:
Could this change be made in Stream.cs - https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L838?

Reply:
https://github.com/dotnet/runtime/pull/35145#discussion_r411573361